src/Corax/Querying/Matches/PharseMatch.cs

```diff
@@ -0,0 +1,201 @@
+    private ByteString _memory;
+    private int MemorySize => _memory.Length / ( sizeof(int) + sizeof(long));
+    
+    
+    private TInner _inner;
+    private readonly FieldMetadata _fieldMetadata;
+    private readonly IndexSearcher _indexSearcher;
+    private readonly ByteString _subsequence;
+    private readonly long _vectorRootPage;
+    private readonly long _rootPage;
+
+
+    public PhraseMatch(in FieldMetadata fieldMetadata, IndexSearcher indexSearcher, TInner inner, ByteString subsequence, long vectorRootPage, long rootPage)
+    {
+        _fieldMetadata = fieldMetadata;
+        _indexSearcher = indexSearcher;
+        _inner = inner;
+        _subsequence = subsequence;
+        _vectorRootPage = vectorRootPage;
+        _rootPage = rootPage;
+        Debug.Assert(_subsequence.Length % sizeof(long) == 0, "this._subsequence.Length % sizeof(long) == 0");
+
+        _memoryHandler = null;
+        _memory = default;
+    }
+
+    public long Count => _inner.Count;
+    public SkipSortingResult AttemptToSkipSorting()
+    {
+        //Filter only, not changing order.
+        return _inner.AttemptToSkipSorting();
+    }
+
+    public QueryCountConfidence Confidence => QueryCountConfidence.Normal;
+    public bool IsBoosting => _inner.IsBoosting;
+    public int Fill(Span<long> matches)
+    {
+        ref var match = ref _inner;
+        var results = match.Fill(matches);
+        return ScanDocumentsTermsEntries(matches.Slice(0, results));
+    }
+
+    public int AndWith(Span<long> buffer, int matches)
+    {
+        var results = _inner.AndWith(buffer, matches);
+        return ScanDocumentsTermsEntries(buffer.Slice(0, results));
+    }
+    
+    private int ScanDocumentsTermsEntries(Span<long> matches)
+    {
+        int currentId = 0;
+        Page p = default;
+        var sequenceToFind = _subsequence.ToSpan<long>();
+        
+        Span<long> buffer = _memoryHandler is null 
+            ? stackalloc long[128] 
+            :  MemoryMarshal.Cast<byte, long>(_memory.ToSpan().Slice(0, sizeof(long) * MemorySize));
+
+        Span<int> indexes = _memoryHandler is null 
+          ? stackalloc int[128]
+          : MemoryMarshal.Cast<byte, int>(_memory.ToSpan().Slice(MemorySize * sizeof(long)));
+        
+        for (var processingId = 0; processingId < matches.Length; ++processingId)
+        {
+            var entryTermsReader = _indexSearcher.GetEntryTermsReader(matches[processingId], ref p);
+            if (entryTermsReader.FindNextStored(_vectorRootPage) == false)
+                continue;
+            
+            //This is value from storage, is not changed since we're seeking to non-stored-value
+            var storedValue = entryTermsReader.StoredField.Value.ToSpan();
+            entryTermsReader.Reset();
+
+            int position = 0;
+            int currentTerm = 0;
+            long termId = -1;
+            while (position < entryTermsReader.StoredField.Value.Length)
```

Comment:
You just reset the `entryTermsReader`, why are you reading the `StoredField` here?
Shouldn't you use `storedValue` instead?